ext/AbstractFFTsChainRulesCoreExt.jl

```diff
@@ -30,14 +30,15 @@ function ChainRulesCore.rrule(::typeof(rfft), x::AbstractArray{<:Real}, dims)
     halfdim = first(dims)
     d = size(x, halfdim)
     n = size(y, halfdim)
-    scale = reshape(
-        [i == 1 || (i == n && 2 * (i - 1) == d) ? 1 : 2 for i in 1:n],
-        ntuple(i -> i == first(dims) ? n : 1, Val(ndims(x))),
-    )
 
     project_x = ChainRulesCore.ProjectTo(x)
     function rfft_pullback(ȳ)
-        x̄ = project_x(brfft(ChainRulesCore.unthunk(ȳ) ./ scale, d, dims))
+        dY = ChainRulesCore.unthunk(ȳ) ./ 2
+        selectdim(dY, halfdim, 1) .*= 2
+        if 2 * (n - 1) == d
+            selectdim(dY, halfdim, n) .*= 2
+        end
```

Comment:
This seems both inefficient (many unnecessary computations) and it breaks non-mutating arrays.

Reply:
Good point. Any suggestion? I used to do the code block below in the previous commit but @gaurav-arya also made a good point that this type conversion might be slow in certain cases.

``` julia
scale = typeof(y)(reshape(
        [i == 1 || (i == n && 2 * (i - 1) == d) ? 1 : 2 for i in 1:n],
        ntuple(i -> i == first(dims) ? n : 1, Val(ndims(x))),
    ))
```